src/mscorlib/shared/System/Threading/ReaderWriterLockSlim.cs

```diff
@@ -984,8 +1062,20 @@ private void ExitAndWakeUpAppropriateWaitersPreferringWriters()
             }
             else if (readercount == 0 && _numWriteWaiters > 0)
             {
+                // Check if a waiter of the same type has already been signaled but hasn't woken yet. If so, avoid signaling
+                // and waking another waiter unnecessarily.
+                WaiterStates signaled = _waiterStates & WaiterStates.WriteWaiterSignaled;
+                if (signaled == WaiterStates.None)
+                {
+                    _waiterStates |= WaiterStates.WriteWaiterSignaled;
+                }
+
                 ExitMyLock();      // Exit before signaling to improve efficiency (wakee will need the lock)
-                _writeEvent.Set();   // release one writer. 
+
+                if (signaled == WaiterStates.None)
+                {
+                    _writeEvent.Set();   // release one writer. 
```

Comment:
But when porting, on the signaling side maybe I should put the state change and call to set together in a finally block. It would prevent abort during ExitMyLock but I'm thinking that should be ok. Or can just set the event before releasing the lock.